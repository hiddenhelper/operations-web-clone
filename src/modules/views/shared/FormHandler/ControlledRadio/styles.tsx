import { makeStyles } from '@material-ui/core/styles';
import { STYLE } from '../../../../../constants';
import { toREM } from '../../../../../utils/generalUtils';

export const useStyles = makeStyles(() => ({
  controlledRadioLabel: {
    color: STYLE.COLOR.SECONDARY_DARKER,
    fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH)}`,
    maxHeight: '20px',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    lineHeight: '24px',
  },
  controlledRadioContainer: {
    marginTop: 15,
    '& .MuiFormControlLabel-label': {
      fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH_BIG)}`,
      color: STYLE.COLOR.SECONDARY_DARKER,
      lineHeight: '27px',
      '&.Mui-disabled': {
        opacity: 0.38,
        '> input[type="radio"] > span': {
          backgroundColor: 'red',
        },
      },
    },

    '& .MuiFormLabel-root.Mui-focused': {
      color: STYLE.COLOR.SECONDARY_DARKER,
    },
  },
  root: {
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
  icon: {
    borderRadius: '50%',
    width: 18,
    height: 18,
    boxShadow: 'inset 0 0 0 1px rgba(16,22,26,.2), inset 0 -1px 0 rgba(16,22,26,.1)',
    backgroundColor: STYLE.COLOR.WHITE,
    backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))',

    '$root.Mui-focusVisible &': {
      outline: `2px auto ${STYLE.COLOR.ACCENT_PRIMARY}`,
      outlineOffset: 2,
    },
    'input:hover ~ &': {
      backgroundColor: '#ebf1f5',
    },
  },
  activeIcon: {
    'input:disabled ~ &': {
      boxShadow: 'none',
    },
  },
  unactiveIcon: {
    'input:disabled ~ &': {
      background: '#fff',
    },
  },
  checkedIcon: {
    backgroundColor: STYLE.COLOR.ACCENT_PRIMARY,
    backgroundImage: 'linear-gradient(180deg,hsla(0,0%,100%,.1),hsla(0,0%,100%,0))',
    '&:before': {
      display: 'block',
      width: 18,
      height: 18,
      backgroundImage: 'radial-gradient(#fff,#fff 28%,transparent 32%)',
      content: '""',
    },
    'input:hover ~ &': {
      backgroundColor: STYLE.COLOR.ACCENT_PRIMARY,
    },
  },
  disabledIcon: {
    background: 'rgba(206,217,224,.5)',
    boxShadow: 'none',
    'input:hover ~ &': {
      backgroundColor: 'rgba(206,217,224,.5)',
    },
  },
}));
