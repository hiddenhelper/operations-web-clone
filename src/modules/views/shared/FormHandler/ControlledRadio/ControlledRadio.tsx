import React from 'react';

import FormControl from '@material-ui/core/FormControl';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormLabel from '@material-ui/core/FormLabel';
import FormControlLabel from '@material-ui/core/FormControlLabel';

import { getDefaultValue } from '../../../../../utils/generalUtils';
import { useStyles } from './styles';

type RadioItem = {
  value: number | string;
  label: string;
};

export interface IControlledRadioProps {
  radioItems: RadioItem[];
  row?: boolean;
  styleClass?: any;
  containerStyleClass?: any;
  disabled?: boolean;
  formControlProps: {
    name: string;
    label: string;
    value: number | string;
    onChange: (data: any) => void;
  };
}

const ControlledRadio = ({ radioItems, row = false, styleClass, disabled = false, containerStyleClass, formControlProps }: IControlledRadioProps) => {
  const classes = useStyles();
  const labelId = `radio-label-${formControlProps.name}`;

  return (
    <div className={`${getDefaultValue(containerStyleClass, '')} ${classes.controlledRadioContainer} controlledRadioContainer`} data-testid="controlled-radio">
      <FormControl component="fieldset" className={styleClass}>
        <FormLabel component="legend" id={labelId} className={classes.controlledRadioLabel}>
          {formControlProps.label}
        </FormLabel>
        <RadioGroup
          aria-labelledby={labelId}
          name={formControlProps.name}
          value={formControlProps.value}
          onChange={formControlProps.onChange}
          row={row}
          data-testid="controlled-radio-group"
        >
          {radioItems.map((item, index) => (
            <FormControlLabel
              key={index}
              value={item.value}
              control={
                <Radio
                  className={classes.root}
                  disableRipple={true}
                  color="default"
                  checkedIcon={<span className={`${classes.icon} ${classes.checkedIcon} ${disabled ? classes.disabledIcon : ''}`} />}
                  icon={<span className={`${classes.icon} ${formControlProps.value === item.value ? classes.activeIcon : classes.unactiveIcon}`} />}
                  inputProps={{ 'data-testid': 'controlled-radio-button' } as any}
                  disabled={disabled}
                />
              }
              label={item.label}
              disabled={disabled}
            />
          ))}
        </RadioGroup>
      </FormControl>
    </div>
  );
};

export default ControlledRadio;
