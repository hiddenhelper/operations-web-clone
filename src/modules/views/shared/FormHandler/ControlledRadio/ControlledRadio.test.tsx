import React from 'react';
import { act, fireEvent, render, RenderResult } from '@testing-library/react';

import ControlledRadio, { IControlledRadioProps } from './ControlledRadio';

describe('Controlled Radio', () => {
  let wrapper: RenderResult;
  let props: IControlledRadioProps = {
    radioItems: [
      {
        value: 'item1',
        label: 'Item 1',
      },
      {
        value: 'item2',
        label: 'Item 2',
      },
    ],
    formControlProps: {
      name: 'Test',
      label: 'Test',
      value: 'item1',
      onChange: jest.fn(),
    },
  };

  it('should render', () => {
    wrapper = render(<ControlledRadio {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should handle item change', async () => {
    wrapper = render(<ControlledRadio {...props} />);

    await act(async () => {
      fireEvent.click(wrapper.getAllByTestId('controlled-radio-button')[1]);
    });

    expect(props.formControlProps.onChange).toHaveBeenCalled();
  });
});
