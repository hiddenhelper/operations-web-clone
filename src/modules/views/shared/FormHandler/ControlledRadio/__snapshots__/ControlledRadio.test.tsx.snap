// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled Radio should render 1`] = `
<body>
  <div>
    <div
      class=" makeStyles-controlledRadioContainer-2 controlledRadioContainer"
      data-testid="controlled-radio"
    >
      <fieldset
        class="MuiFormControl-root"
      >
        <legend
          class="MuiFormLabel-root makeStyles-controlledRadioLabel-1"
          id="radio-label-Test"
        >
          Test
        </legend>
        <div
          aria-labelledby="radio-label-Test"
          class="MuiFormGroup-root"
          data-testid="controlled-radio-group"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiRadio-root makeStyles-root-3 PrivateSwitchBase-checked-10 Mui-checked"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-12"
                  data-testid="controlled-radio-button"
                  name="Test"
                  type="radio"
                  value="item1"
                />
                <span
                  class="makeStyles-icon-4 makeStyles-checkedIcon-7 "
                />
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Item 1
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiRadio-root makeStyles-root-3"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-12"
                  data-testid="controlled-radio-button"
                  name="Test"
                  type="radio"
                  value="item2"
                />
                <span
                  class="makeStyles-icon-4 makeStyles-unactiveIcon-6"
                />
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Item 2
            </span>
          </label>
        </div>
      </fieldset>
    </div>
  </div>
</body>
`;
